File: solutions/fsharp/protein-translation/ProteinTranslation.fs
module ProteinTranslation

let toPolypeptide =
    function
    | "AUG" -> "Methionine"
    | "UUC"
    | "UUU" -> "Phenylalanine"
    | "UUA"
    | "UUG" -> "Leucine"
    | "UCU"
    | "UCC"
    | "UCA"
    | "UCG" -> "Serine"
    | "UAU"
    | "UAC" -> "Tyrosine"
    | "UGC"
    | "UGU" -> "Cysteine"
    | "UGG" -> "Tryptophan"
    | _ -> "STOP"

let proteins (rna: string) =
    Seq.chunkBySize 3 rna
    |> Seq.map (System.String >> toPolypeptide)
    |> Seq.takeWhile ((<>) "STOP")
    |> Seq.toList

File: solutions/scala/protein-translation/src/main/scala/ProteinTranslation.scala
object ProteinTranslation:
  private val Stop = "STOP"
  private type RnaSequence = String
  private type Protein = String
  private type Codon = String

  private def codonToProtein(codon: Codon): Protein =
    codon match
      case "AUG" => "Methionine"
      case "UUU" | "UUC" => "Phenylalanine"
      case "UUA" | "UUG" => "Leucine"
      case "UCU" | "UCC" | "UCA" | "UCG" => "Serine"
      case "UAU" | "UAC" => "Tyrosine"
      case "UGU" | "UGC" => "Cysteine"
      case "UGG" => "Tryptophan"
      case "UAA" | "UAG" | "UGA" => Stop
      case invalid => throw new IllegalArgumentException(s"Invalid codon: $invalid")

  def proteins(rna: RnaSequence): Seq[Protein] =
    rna.grouped(3).map(codonToProtein).takeWhile(_ != Stop).toSeq
